kstatus/status/generic.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2019 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package status
+
+import (
+	"fmt"
+
+	corev1 "sigs.k8s.io/kustomize/pseudo/k8s/api/core/v1"
+	"sigs.k8s.io/kustomize/pseudo/k8s/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func checkGenericProperties(u *unstructured.Unstructured) (*Result, error) {
+	obj := u.UnstructuredContent()
+
+	// Check if the resource is scheduled for deletion
+	deletionTimestamp, found, err := unstructured.NestedString(obj, "metadata", "deletionTimestamp")
+	if err != nil {
+		return nil, fmt.Errorf("error looking up metadata.deletionTimestamp from resource: %v", err)
+	}
+	if found && deletionTimestamp != "" {
+		return &Result{
+			Status:     TerminatingStatus,
+			Message:    "Resource scheduled for deletion",
+			Conditions: []Condition{},
+		}, nil
+	}
+
+	// ensure that the meta generation is observed
+	generation, found, err := unstructured.NestedInt64(u.Object, "metadata", "generation")
+	if err != nil {
+		return nil, fmt.Errorf("error looking up metadata.generation from resource: %v", err)
+	}
+	if !found {
+		return nil, fmt.Errorf("unable to find metadata.generation from resource")
+	}
+	observedGeneration, found, err := unstructured.NestedInt64(u.Object, "status", "observedGeneration")
+	if err != nil {
+		return nil, fmt.Errorf("error looking up status.observedGeneration from resource: %v", err)
+	}
+	if found {
+		// Resource does not have this field, so we can't do this check.
+		// TODO(mortent): Verify behavior of not set vs does not exist.
+		if observedGeneration != generation {
+			message := fmt.Sprintf("%s generation is %d, but latest observed generation is %d", u.GetKind(), generation, observedGeneration)
+			return &Result{
+				Status:     InProgressStatus,
+				Message:    message,
+				Conditions: []Condition{newInProgressCondition("LatestGenerationNotObserved", message)},
+			}, nil
+		}
+	}
+
+	// Check if the resource has any of the standard conditions. If so, we just use them
+	// and no need to look at anything else.
+	objWithConditions, err := GetObjectWithConditions(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, cond := range objWithConditions.Status.Conditions {
+		if cond.Type == string(ConditionInProgress) && cond.Status == corev1.ConditionTrue {
+			return newInProgressStatus(cond.Reason, cond.Message), nil
+		}
+		if cond.Type == string(ConditionFailed) && cond.Status == corev1.ConditionTrue {
+			return &Result{
+				Status:  FailedStatus,
+				Message: cond.Message,
+				Conditions: []Condition{
+					{
+						Type:    ConditionFailed,
+						Status:  corev1.ConditionTrue,
+						Reason:  cond.Reason,
+						Message: cond.Message,
+					},
+				},
+			}, nil
+		}
+	}
+
+	return nil, nil
```

Comment:
Are the Current and Terminating conditions accounted for?

Reply:
I don't think we should have a condition for Terminating. There is already a clear way to check if a resource is in the terminating state, and that is by looking at the metadata.deletionTimestamp field.

Since we are going with the "abnormal true" approach, we don't have a condition for Current. The absence of any "abnormal" conditions (like InProgress or Failed/Stalled) means it is Current. This is a challenge here since we have no way of telling if a resource have no abnormal conditions because it is Current or if it has no abnormal conditions because it doesn't set them at all. This currently works because we eventually fall back on saying a resource is current if we can't find any "evidence" otherwise.